aip/general/0128.md

```diff
@@ -0,0 +1,106 @@
+principle of "configuration as code": the user specifies the complete intended
+landscape, and tooling is responsible for making whatever changes are necessary
+to achieve the user's specification.
+
+These tools are **declarative**: rather than specifying specific _actions_ to
+take, they specify the desired _outcome_, with the actions being derived based
+on the differences between the current landscape and the intended one.
+
+Furthermore, there are numerous popular DevOps tools, with more being
+introduced each year. Integrating hundreds of resource types with multiple
+tools requires strong consistency, so that integration can be automated.
+
+## Guidance
+
+Services **should** clearly delineate between "control plane" operations and
+"data plane" operations, ideally through the use of distinct services with
+their own interface definition documents.
+
+- Control plane operations are responsible for managing the _lifecycle_ of
+  resources.
+- Data plane operations are responsible for managing the _content_ of
+  resources.
+
+The same resource **may** have both control plane operations and data plane
+operations associated with it. For example, a database API would have
+operations to create or delete database tables (control plane) as well as
+operations to write and read rows to that table (data plane).
+
+### Resources
+
+Resources that are declarative-friendly **must** use only strongly-consistent
+standard methods for managing resource lifecycle, which allows tools to support
+these resources generically, as well as conforming to other
+declarative-friendly guidance (see [further reading](#further-reading)).
+
+Declarative-friendly resources **should** designate that they follow the
+declarative-friendly style:
+
+```proto
+message Book {
+  option (google.api.resource) = {
+    type: "library.googleapis.com/Book"
+    pattern: "publishers/{publisher}/books/{book}"
+    style: DECLARATIVE_FRIENDLY
+  };
+
+  string name = 1;
+  // Other fields...
+}
+```
+
+### Annotations
+
+Declarative-friendly resources **must** include a
+`map<string, string> annotations` field to allow clients to store small amounts
+of arbitrary data.
+
+The `annotations` field **must** use the [Kubernetes limits][] to maintain wire
+compatibility, and **should** require dot-namespaced annotation keys to prevent
+tools from trampling over one another.
+
+**Note:** Annotations are distinct from various forms of labels. Labels can be
+used by server-side policies, such as IAM conditions. Annotations exist to
+allow client tools to store their own state information without requiring a
+database.
+
+<!-- prettier-ignore -->
+[kubernetes limits]: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set
+
+### Reconciliation
+
+If a resource takes time (more than a few seconds) for updates to be realized,
+the service **should** include an output only `bool reconciling` field to
```

Comment:
I'd hyphenate output-only but I can't remember whether we've already disagreed about hyphenation styles. I'm certainly not going to push it :)

Reply:
We have generally not done it, but we also generally do not use it as an adjective like this, and I agree that output-only here is correct. Choosing to rephrase the paragraph to be consistent.

Reply:
Done.